quickwit-index/src/actors/indexer.rs

```diff
@@ -0,0 +1,364 @@
+}
+
+struct ImmutableState {
+    index_id: String,
+    index_config: Arc<dyn IndexConfig>,
+    commit_policy: CommitPolicy, //< TODO should move into the index config.
+    // splits index writer will write in TempDir within this directory
+    indexing_scratch_directory: ScratchDirectory,
+    timestamp_field_opt: Option<Field>,
+}
+
+impl ImmutableState {
+    fn create_indexed_split(&self) -> anyhow::Result<IndexedSplit> {
+        let schema = self.index_config.schema();
+        let indexed_split = IndexedSplit::new_in_dir(
+            self.index_id.clone(),
+            &self.indexing_scratch_directory,
+            schema,
+        )?;
+        Ok(indexed_split)
+    }
+
+    /// Returns the current_indexed_split. If this is the first message, then
+    /// the indexed_split does not exist yet.
+    ///
+    /// This function will then create it, and can hence return an Error.
+    fn get_or_create_current_indexed_split<'a>(
+        &self,
+        current_split_opt: &'a mut Option<IndexedSplit>,
+        ctx: &ActorContext<Indexer>,
+    ) -> anyhow::Result<&'a mut IndexedSplit> {
+        if current_split_opt.is_none() {
+            let new_indexed_split = self.create_indexed_split()?;
+            let commit_timeout = IndexerMessage::CommitTimeout {
+                split_id: new_indexed_split.split_id.clone(),
+            };
+            ctx.schedule_self_msg_blocking(self.commit_policy.timeout, commit_timeout);
+            *current_split_opt = Some(new_indexed_split);
+        }
+        let current_index_split = current_split_opt.as_mut().with_context(|| {
+            "No index writer available. Please report: this should never happen."
+        })?;
+        Ok(current_index_split)
+    }
+
+    fn parse_doc(&self, doc_json: &str, counters: &mut IndexerCounters) -> Option<Document> {
+        let doc_parsing_result = self.index_config.doc_from_json(&doc_json);
+        match doc_parsing_result {
+            Ok(doc) => Some(doc),
+            Err(doc_parsing_error) => {
+                // TODO we should at least keep track of the number of parse error.
+                warn!(err=?doc_parsing_error);
+                counters.num_parse_errors += 1;
+                None
+            }
+        }
+    }
+
+    fn process_batch(
+        &self,
+        batch: RawDocBatch,
+        current_split_opt: &mut Option<IndexedSplit>,
+        counters: &mut IndexerCounters,
+        ctx: &ActorContext<Indexer>,
+    ) -> Result<(), ActorTermination> {
+        let indexed_split = self.get_or_create_current_indexed_split(current_split_opt, ctx)?;
+        for doc_json in batch.docs {
+            indexed_split.size_in_bytes += doc_json.len() as u64;
+            let doc = if let Some(doc) = self.parse_doc(&doc_json, counters) {
+                doc
+            } else {
+                continue;
+            };
+            if let Some(timestamp) = extract_timestamp_if_any(&doc, self.timestamp_field_opt) {
```

Comment:
note: if we have a timestamp field on the index, it could be nice to fail if the field is not the document.

Reply:
Agreed and fixed.